network/server.go

```diff
@@ -327,29 +329,63 @@ func (s *Server) setupBootnodes() error {
 }
 
 // keepAliveMinimumPeerConnections will attempt to make new connections
-// if the active peer count is lesser than the specified limit.
+// if the active peer count is lesser than the specified limit. And will attempt
+// to ensure that all bootnodes are connected to
 func (s *Server) keepAliveMinimumPeerConnections() {
+	bootNodes := s.bootnodes.getBootnodes()
+	ourId := s.host.ID()
+
+	// remove ourselves from the list of bootNodes if present
+	for i, bootNode := range bootNodes {
+		if bootNode.ID == ourId {
+			bootNodes = append(bootNodes[:i], bootNodes[i+1:]...)
+		}
+	}
+
+	s.logger.Info("[serverKeepAlive] starting keepalive", "bootNodes", bootNodes)
+
 	for {
 		select {
 		case <-time.After(10 * time.Second):
 		case <-s.closeCh:
 			return
 		}
 
-		if s.numPeers() < MinimumPeerConnections {
-			if s.config.NoDiscover || !s.bootnodes.hasBootnodes() {
-				// dial unconnected peer
-				randPeer := s.GetRandomPeer()
-				if randPeer != nil && !s.IsConnected(*randPeer) {
-					s.addToDialQueue(s.GetPeerInfo(*randPeer), common.PriorityRandomDial)
+		currentPeers := s.numPeers()
+
+		// if we're in a setup where we just want random peering
+		if currentPeers < MinimumPeerConnections && (s.config.NoDiscover || !s.bootnodes.hasBootnodes()) {
+			// dial unconnected peer
+			randPeer := s.GetRandomPeer()
+			s.logger.Info("[serverKeepAlive] adding random peer", "id", randPeer.String())
+			if randPeer != nil && !s.IsConnected(*randPeer) {
+				s.addToDialQueue(s.GetPeerInfo(*randPeer), common.PriorityRandomDial)
+			}
+		}
+
+		// ensure we attempt to connect to all of our bootNodes
+		peers := s.Peers()
+		var added []string
```

Comment:
Optional change:
```suggestion
		added := make([]string, 0, len(bootNodes))
```

Reply:
I opted for using `MaximumDialsPerKeepAlive` as the capacity as we won't ever add more than this.